### PR TITLE
[Backend] Add clearSubtargetMap API for TargetMachine.

### PR DESCRIPTION
**Pull request by weiweichen**
- [x] Add `clearSubtargetInfo` API to TargetMachine and each backend to make it possible to release memory used in each backend's `SubtargetInfo` map if needed. Keep this API as `protected` so that it will be used with precautions.

### CODE REVIEW

**Review by arsenm — COMMENTED** [review]
I think this is too specific. What is the use case for this? If you're tearing down the TargetMachine, I would expect to release everything. If you're reusing the TargetMachine, I would also expect to just leave everything as it is

**Comment by weiweichen** [comment]
> I think this is too specific. What is the use case for this? If you're tearing down the TargetMachine, I would expect to release everything. If you're reusing the TargetMachine, I would also expect to just leave everything as it is

I don't want to tear down the whole, I just want to release the memory in this map specifically and dedup them into a centralized place to reduce memory footprint when I have multiple TargetMachines running in parallel in the system.

**Comment by arsenm** [comment]
> I don't want to tear down the whole TargetMachine, I just want to release the memory in this map specifically and dedup them into a centralized place to reduce memory footprint when I have multiple TargetMachines running in parallel in the system.

Why do you need multiple TargetMachines from the same target to compile in parallel?

**Comment by weiweichen** [comment]
> > I don't want to tear down the whole TargetMachine, I just want to release the memory in this map specifically and dedup them into a centralized place to reduce memory footprint when I have multiple TargetMachines running in parallel in the system.
> 
> Why do you need multiple TargetMachines from the same target to compile in parallel?

To parallelize llvm compilation for more efficient compilation time.

**Comment by weiweichen** [comment]
> > To parallelize llvm compilation for more efficient compilation time.
> 
> But I would expect you can use one machine for different compilations. Is the issue TargetOptions?

Not exactly. It's the that `TargetMachine` (like LLVMContext and MCContext) cannot be shared among parallel llc pipelines because it's not read-only and not thread-safe. `SubtargetInfo` is part of `TargetMachine` (or rather `LLVMTargetMachine`), it is a StringMap to dedup those can be shared among MachineFunctions that uses the same `TargetMachine`, but not when there are multiple `TargetMachine`s. This is not mainly a functionality issue but memory footprint during compilation time and can potentially reduced where there are many parallel pipelines running.

**Comment by arsenm** [comment]
If we had an explicit initialization of all subtargets / TargetOptions needed by a module, and made the contexts external (i.e. #111234), can you just use the one TargetMachine? Other than those, the TargetMachine should be stateless

**Comment by weiweichen** [comment]
> If we had an explicit initialization of all subtargets / TargetOptions needed by a module, and made the contexts external (i.e. #111234), can you just use the one TargetMachine? Other than those, the TargetMachine should be stateless

Yes, it would solve the problem if the initialization can be don explicitly once and the `TargetMachine` is stateless. Thank for pointing me to the PR! I'll take a detailed look at the it. Though one thing I notice that the PR is not changing the [`mutable StringMap<std::unique_ptr<AArch64Subtarget>> SubtargetMap;`](https://github.com/llvm/llvm-project/pull/111234/files#diff-f7cbbcee7ebd578cd6b1a442bfe7ec533385b783deb2dc4e964adfdb5ca2eed5R27) private/protected member in concrete TargetMachines for different backends. They are `mutable` and being modified mostly in the same pattern [here](https://github.com/llvm/llvm-project/blob/b0a25468faca32d9db4d35e56fb120ed4eaeec09/llvm/lib/Target/AArch64/AArch64TargetMachine.cpp#L467). Is there a plan to change that to make TargetMachine fully stateless as well?

**Comment by arsenm** [comment]
> They are mutable and being modified mostly in the same pattern [here](https://github.com/llvm/llvm-project/blob/b0a25468faca32d9db4d35e56fb120ed4eaeec09/llvm/lib/Target/AArch64/AArch64TargetMachine.cpp#L467). Is there a plan to change that to make TargetMachine fully stateless as well?

No, I'm not aware of one. The other problematic piece is TargetOptions

**Comment by npanchen** [comment]
> If we had an explicit initialization of all subtargets / TargetOptions needed by a module, and made the contexts external (i.e. #111234),

If by "explicit initialization" you refer to calling `getSubtargetImpl`, this won't be sufficient, unfortunately. Having something like `TargetMachine::setSubtarget(shared_ptr<TargetSubtargetInfo>)` might help.

>  can you just use the one TargetMachine? Other than those, the TargetMachine should be stateless

To describe the problem a bit more: there's parallel compilation that use identical TargetMachines, but different instances. it's known that at some point `TargetSubtargetInfo` can be released to reduce memory footprint.  Ideally, if TargetMachine is stateless, there will be no need to have different instances.

if you or @topperc can recommend what can be done in a short term, that will be great.

**Comment by arsenm** [comment]
> If by "explicit initialization" you refer to calling `getSubtargetImpl`, this won't be sufficient, unfortunately. Having something like `TargetMachine::setSubtarget(shared_ptr<TargetSubtargetInfo>)` might help.

I mean the lazy construction of subtargets is an issue. You would have to pre-populate the map with any subtargets that are needed in the module (i.e. loop through every function and query the subtarget for each)

**Comment by npanchen** [comment]
> I mean the lazy construction of subtargets is an issue. 

I don't think lazy construction itself is an issue here: regardless if it's lazy or not, multiple instance of a target machine is going to keep subtargets alive, thus memory footprint is going to stay high for longer than needed.

> You would have to pre-populate the map with any subtargets that are needed in the module (i.e. loop through every function and query the subtarget for each)

That's absolutely fine with a suggestion I made. However, I don't want to derail conversation about that patch. Does it look fine approach to be taken ?

**Comment by arsenm** [comment]
> > I mean the lazy construction of subtargets is an issue.
> 
> I don't think lazy construction itself is an issue here: regardless if it's lazy or not, multiple instance of a target machine is going to keep subtargets alive, thus memory footprint is going to stay high for longer than needed.

It would be for parallelization. What I am advocating for is making so you can do parallel compilations with a single TargetMachine instance, not one per thread. 

Is the memory footprint of subtargets really so large? I thought they were pretty tiny

**Comment by weiweichen** [comment]
> > > I mean the lazy construction of subtargets is an issue.
> > 
> > 
> > I don't think lazy construction itself is an issue here: regardless if it's lazy or not, multiple instance of a target machine is going to keep subtargets alive, thus memory footprint is going to stay high for longer than needed.
> 
> It would be for parallelization. What I am advocating for is making it so you can do parallel compilations with a single TargetMachine instance, not one per thread.
> 
> Is the memory footprint of subtargets really so large? I thought they were pretty tiny

They were pretty tiny by themselves (~4M for example), but they accumulate/multiply when there are many parallel tasks (for higher-level of parallelization) and we need a barrier to collect the codegen result together before doing the final AsmPrint. We see about 60% peak memory reduce if we can release these SubtargetInfos earlier and replace them with a shared copy.

**Comment by weiweichen** [comment]
Making the API name more general.

**Comment by weiweichen** [comment]
Could someone take look at this PR when you have sometime? If there is no more objections in a few days, I'll merge the PR and will definitely make followups or revert if needed. Thank you!

File: llvm/include/llvm/Target/TargetMachine.h [hunk 1]
```diff
@@ -448,6 +448,9 @@ class LLVMTargetMachine : public TargetMachine {
 
   void initAsmInfo();
 
+  /// Reset internal state.
+  virtual void reset() {};
```
Comment (topperc) [inline llvm/include/llvm/Target/TargetMachine.h:452]: Should this be pure virtual since every target needs to implement this?

**Review by tjk213 — COMMENTED** [review]
I like the name-change to a general API rather than exposing the internal detail of `SubtargetMap`. However, with a general name like `reset()` I would expect the function to reset *all* internal state, but I'm not sure that this PR accomplishes that for all targets. We could consider raising an unimplemented error in the base class and only adding `reset()` to the targets that you need to be a little more defensive about it.

Other than that, this looks good to me.

**Comment by weiweichen** [comment]
> I like the name-change to a general API rather than exposing the internal detail of `SubtargetMap`. However, with a general name like `reset()` I would expect the function to reset _all_ internal state, but I'm not sure that this PR accomplishes that for all targets. We could consider raising an unimplemented error in the base class and only adding `reset()` to the targets that you need to be a little more defensive about it.
> 
> Other than that, this looks good to me.

I got other reviewer's request to make the name of the function more general although the intent is to just clear the one data structure. I'm not sure how to balance that comment with yours though I agree it makes more sense to me that `reset()` resets _all_ internal state. 

I'll keep the general name of the API and remove the implementation for the targets I don't need for now to keep it as a no-op.

File: llvm/include/llvm/Target/TargetMachine.h (hunk 1 above)
Comment (weiweichen) [reply llvm/include/llvm/Target/TargetMachine.h:452]: This is a good point! Trying to balance all the review comments I got and I think I'll remove the implementation from the backend that I don't particularly care about and leave them with the default implementation here as a no-op and can always overload with specific implementations should the need arises in the future.

**Pull request merged**